packages/ledgerstate/transaction.go

```diff
@@ -0,0 +1,895 @@
+		unlockBlocks: unlockBlocks,
+	}
+}
+
+// TransactionFromBytes unmarshals an Transaction from a sequence of bytes.
+func TransactionFromBytes(bytes []byte) (transaction *Transaction, consumedBytes int, err error) {
+	marshalUtil := marshalutil.New(bytes)
+	if transaction, err = TransactionFromMarshalUtil(marshalUtil); err != nil {
+		err = xerrors.Errorf("failed to parse Transaction from MarshalUtil: %w", err)
+		return
+	}
+	consumedBytes = marshalUtil.ReadOffset()
+
+	return
+}
+
+// TransactionFromMarshalUtil unmarshals a Transaction using a MarshalUtil (for easier unmarshaling).
+func TransactionFromMarshalUtil(marshalUtil *marshalutil.MarshalUtil) (transaction *Transaction, err error) {
+	readStartOffset := marshalUtil.ReadOffset()
+
+	payloadSize, err := marshalUtil.ReadUint32()
+	if err != nil {
+		err = xerrors.Errorf("failed to parse payload size from MarshalUtil (%v): %w", err, cerrors.ErrParseBytesFailed)
+		return
+	}
+	// a payloadSize of 0 indicates the payload is omitted and the payload is nil
+	if payloadSize == 0 {
+		return
+	}
+	payloadType, err := payload.TypeFromMarshalUtil(marshalUtil)
+	if err != nil {
+		err = xerrors.Errorf("failed to parse payload Type from MarshalUtil: %w", err)
+		return
+	}
+	if payloadType != TransactionType {
+		err = xerrors.Errorf("payload type '%s' does not match expected '%s': %w", payloadType, TransactionType, cerrors.ErrParseBytesFailed)
+		return
+	}
+
+	transaction = &Transaction{}
+	if transaction.essence, err = TransactionEssenceFromMarshalUtil(marshalUtil); err != nil {
+		err = xerrors.Errorf("failed to parse TransactionEssence from MarshalUtil: %w", err)
+		return
+	}
+	if transaction.unlockBlocks, err = UnlockBlocksFromMarshalUtil(marshalUtil); err != nil {
+		err = xerrors.Errorf("failed to parse UnlockBlocks from MarshalUtil: %w", err)
+		return
+	}
+
+	parsedBytes := marshalUtil.ReadOffset() - readStartOffset
+	if parsedBytes != int(payloadSize)+marshalutil.UINT32_SIZE {
+		err = xerrors.Errorf("parsed bytes (%d) did not match expected size (%d): %w", parsedBytes, payloadSize, cerrors.ErrParseBytesFailed)
+		return
+	}
+
+	if len(transaction.unlockBlocks) != len(transaction.essence.Inputs()) {
+		err = xerrors.Errorf("amount of UnlockBlocks (%d) does not match amount of Inputs (%d): %w", len(transaction.unlockBlocks), len(transaction.essence.inputs), cerrors.ErrParseBytesFailed)
+		return
+	}
+
+	return
+}
+
+// TransactionFromObjectStorage restores a Transaction that was stored in the ObjectStorage.
+func TransactionFromObjectStorage(key []byte, data []byte) (transaction objectstorage.StorableObject, err error) {
+	if transaction, _, err = TransactionFromBytes(data); err != nil {
+		err = xerrors.Errorf("failed to parse Transaction from bytes: %w", err)
+		return
+	}
+
+	transactionID, _, err := TransactionIDFromBytes(key)
+	if err != nil {
+		err = xerrors.Errorf("failed to parse TransactionID from bytes: %w", err)
+		return
+	}
+	transaction.(*Transaction).id = &transactionID
```

Comment:
Why is the explicit cast necessary? `TransactionFromBytes` returns already `*Transaction`

Reply:
Because the named parameter is of the type objectstorage.StorableObject